Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -678,16 +741,29 @@ TEST_F(SerializerTest, BadFieldValueTagWithOtherValidTagsPresent) {
 
   // Decode the bytes into the model
   Reader reader = Reader::Wrap(bytes.data(), bytes.size());
-  absl::optional<FieldValue> actual_model =
-      serializer.DecodeFieldValue(&reader);
+  google_firestore_v1beta1_Value nanopb_proto =
+      google_firestore_v1beta1_Value_init_zero;
+  reader.ReadNanopbMessage(google_firestore_v1beta1_Value_fields,
+                           &nanopb_proto);
+  FieldValue actual_model = serializer.DecodeFieldValue(&reader, nanopb_proto);
+  reader.FreeNanopbMessage(google_firestore_v1beta1_Value_fields,
+                           &nanopb_proto);
   EXPECT_OK(reader.status());
 
   // Ensure the decoded model is as expected.
   FieldValue expected_model = FieldValue::FromBoolean(true);
-  EXPECT_EQ(FieldValue::Type::Boolean, actual_model->type());
-  EXPECT_EQ(expected_model, *actual_model);
+  EXPECT_EQ(FieldValue::Type::Boolean, actual_model.type());
+  EXPECT_EQ(expected_model, actual_model);
 }
 
+/* TODO(rsgowman): nanopb doesn't handle cases where the type as read on the
+ * wire, and the type as defined by the tag disagree. In these cases, the type
+ * defined by the tag "wins" and the corruption is ignored.
+ *
+ * We need to figure out what, if anything, we want to do about this. For now,
```

Comment:
I think the answer is to do what protobuf does. Happy to take changes that implement that in a separate PR.

Reply:
protobuf fails to decode.

By the time we get the proto from nanopb, it's too late to notice the corruption. We could patch nanopb to notice the corruption and return an error. (Again, it would likely add code size/cycles to nanopb, though it would be fairly trivial.)

Reply:
To my eye this doesn't seem worth pursuing. Yes it's true that this could conceivably happen but nanopb's behavior isn't insane. If you agree I think we can just drop this.

Reply:
dropped.